### ISSUE TITLE
Best way to handle Singleton implementation

### ISSUE THREAD

**Issue by KniKnaKnorke** [JPro-one/JPro-tickets#85]
My JavaFX-Application has some classes which implement the singleton patter. 
As far as I know jPro runs all instances in one JVM, so the singleton instances are shared between the application instances. 

What is the best way to solve this problem? Do I need to refactor my application and remove the singleton pattern?

**Comment by FlorianKirmaier** [comment]
Yes, if your Singleton should be per Application, it should be rewritten not to be a singleton.

There are some alternative approaches, like 1 user per JVM or 1 Classloader per user, but we usually don't use them.
If you are interested in them, feel free to contact us directly under https://www.jpro.one/?page=contact

**Comment by Hugolarson** [comment]
> My JavaFX-Application has some classes which implement the singleton patter.
> As far as I know jPro runs all instances in one JVM, so the singleton instances are shared between the application instances.
> 
> What is the best way to solve this problem? Do I need to refactor my application and remove the singleton pattern?

Had the same problem. Investigate ThreadLocal approach. Had to do some rewriting but not terrible much.

**Comment by Hugolarson** [comment]
Follow up with a question to Florian.
Is there a way to only allow one instance of each browser/client?

**Comment by FlorianKirmaier** [comment]
ThreadLocal often isn't useful, because in JavaFX you are typically using the same Thread, 
If you want to remember users, my recommendation would be to set a cookie for the user and recognize the user based on the cookie.
I agree, that JPro should have something more direct to share variables between instances in one browser.

**Comment by KniKnaKnorke** [comment]
Thanks for the fast response. I will have a look at the mentioned topics here.

**Comment by Hugolarson** [comment]
In the early days I did some test and found out that each connection creates a JavaFx Application thread. I use InheritableThreadLocal and it works. Do you mean that entire JPro server use one and only JavaFx Application thread for all clients?

**Comment by FlorianKirmaier** [comment]
@Hugolarson 
Yes, you must have done something wrong in your test. 
JPro uses only 1 JavaFXThread per server, it wasn't any different in the past.

**Comment by Hugolarson** [comment]
How about storing global variables in mainstage.setUserData(...)?

**Comment by Hugolarson** [comment]
How about Stages. Are they shared also?
I use a lot Window.getWindows() which returns a List of windows and then I check if the window object isFocused(). Is isFocused() reliable? Or can several stages be focused because nature of Jpro?
I think there must be a way to to ask JPro which stage is active with a  static call like WebApi.getActiveStage() in a UI Event.

I was thinking to store global data in the main stage object with stage.setUserData(map)
